### PR TITLE
Refactor: Enhance file handling and code editing functionality

### PR DESCRIPTION
**Pull request by PierrunoYT**
# Enhance file handling and code editing functionality

## What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?

This PR addresses several issues related to file handling, error management, and code editing functionality:

1. Inconsistent error handling in the backend (listen.py)
2. Lack of proper input sanitization and type checking in file operations (fileService.ts)
3. Limited user feedback during file saving operations in the code editor (CodeEditor.tsx)

This PR fixes the open issue #855: "Improve file handling and add save functionality" (https://github.com/OpenDevin/OpenDevin/issues/855). It significantly improves the robustness and user experience of the application's file handling and code editing features.

## Give a brief summary of what the PR does, explaining any non-trivial design decisions

This PR introduces the following key changes:

1. In `listen.py`:
   - Refactored error handling to use FastAPI's `HTTPException` consistently
   - Added a new `/api/save-file` endpoint with proper agent state checks for increased security
   - Simplified WebSocket event handling logic

2. In `fileService.ts`:
   - Implemented comprehensive error handling with try-catch blocks
   - Added input sanitization using `encodeURIComponent()` for file names and paths
   - Introduced type checking for API responses
   - Added a new `saveFile()` function for improved file saving capabilities

3. In `CodeEditor.tsx`:
   - Implemented a new save button with dynamic colors based on save status
   - Added a save notification system and "Last saved" timestamp display
   - Integrated with Redux for better state management of code and agent state

Design decisions:
- Used FastAPI's built-in error handling to maintain consistency and leverage framework capabilities
- Implemented input sanitization to prevent potential security issues with special characters in file names
- Chose to display save status visually in the UI to provide immediate feedback to users
- Utilized Redux for state management to handle complex state interactions efficiently

## Other references

https://github.com/OpenDevin/OpenDevin/assets/95778421/811533ac-0ec1-406d-92e8-b6c800f7d633

- FastAPI documentation on exception handling: https://fastapi.tiangolo.com/tutorial/handling-errors/
- React Redux usage guidelines: https://react-redux.js.org/introduction/getting-started
- OWASP Input Validation Cheat Sheet: https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html

### CODE REVIEW

**Comment by PierrunoYT** [comment]
Right now it's only working when the agent is paused or the task is finished but not when the Agent awaits User Input like Continue

File: opendevin/server/listen.py
Comment (iFurySt) [inline opendevin/server/listen.py:364]: Why do you delete the docstring? 🤔

Comment (PierrunoYT) [reply opendevin/server/listen.py:364]: Yeah sorry it's wrong

Comment (PierrunoYT) [reply opendevin/server/listen.py:364]: @iFurySt I tried to add it back

File: frontend/src/components/CodeEditor.tsx
Comment (tobitege) [inline frontend/src/components/CodeEditor.tsx:153]: Why change this?

File: frontend/src/components/CodeEditor.tsx
Comment (tobitege) [inline frontend/src/components/CodeEditor.tsx:167]: This should be made translatable via i18n in translations.json

File: frontend/src/services/fileService.ts
Comment (tobitege) [inline frontend/src/services/fileService.ts:27]: This comment relates to all added exception handling in this PR!
exception handling should use "toast.Error" as a user feedback in the UI.
"console" will happen behind the scenes.
Add this to imports at top of file:
```ts
import toast from "#/utils/toast";
```

File: frontend/src/components/CodeEditor.tsx
Comment (PierrunoYT) [reply frontend/src/components/CodeEditor.tsx:153]: Not sure tbh

File: frontend/src/services/fileService.ts
Comment (tobitege) [inline frontend/src/services/fileService.ts:45]: While we're at it: some safety checks could be added before try line, like
```ts
  if (!filePath || filePath.includes('..')) {
    toast.error('Invalid file path'); // needs translation too
  }
```

**Review by tobitege — CHANGES_REQUESTED** [review]
Couple of changes needed, ideally.

File: frontend/src/services/fileService.ts
Comment (PierrunoYT) [reply frontend/src/services/fileService.ts:30]: I tried to fix it

File: frontend/src/services/fileService.ts
Comment (PierrunoYT) [reply frontend/src/services/fileService.ts:55]: I tried to fix it

File: opendevin/server/listen.py
Comment (PierrunoYT) [reply opendevin/server/listen.py:82]: Added Docstrings back

File: opendevin/server/listen.py
Comment (PierrunoYT) [reply opendevin/server/listen.py]: I put them each on a single line

**Comment by tobitege** [comment]
I'll try to get this into a working and reviewable state today/tomorrow.

**Comment by tobitege** [comment]
Fixed the PR up:
- eslint / prettier happy (before merge with main)
- Changing file didn't update save button
- translations should be complete now
- restored parts of missing docstrings

Decided to remove the extra display of a save timestamp that made the UI more complicated.

@iFurySt @amanape if you guys would have a look please for actual approval, I'm biased. 🤣

File: frontend/src/services/fileService.ts
Comment (amanape) [inline frontend/src/services/fileService.ts:1]: You are `throw`ing the error in the catch block, where thrown errors are supposed to be caught. This means the error will propagate upwards and potentially stop the app either way. A better thing to do here is to wrap the component that calls `selectFile` in a `try-catch` block and render a `toast` when an error is caught. This also "lightens" the load/responsibility of `selectFile` to just attempting to make an API call, and the rest is handled by the component/function where it is called

Same for the `listFiles` and `saveFile` functions

File: frontend/src/components/file-explorer/CodeEditor.tsx
Comment (amanape) [inline frontend/src/components/file-explorer/CodeEditor.tsx:1]: There is something a little off about this but I can't put my finger on it. The code might be somewhat packed and the UI itself bloated.

Not necessarily a _big_ problem, so if others are OK with it, no problem

**Review by amanape — COMMENTED** [review]
This is a good feature, but I think we need some extra feedback on this decision from @neubig and/or @xingyaoww. I remember some conversation about allowing the user to make edits from the editor, but there may be a need to have OpenDevin be aware of the changes the user makes. Otherwise, it may or may not negatively affect OpenDevin's output (creating a file that already exists that end up overriding it, calling a function that has been renamed, updating an incorrect range of lines that has been formatted, etc...).

Either way, I am pretty sure users have been editing files through their own editor anyways, but since we don't have any form of autocomplete/shortcuts/intelli-sense with these changes, users may still prefer using their own editor unless it is for very simple things.

File: frontend/src/services/fileService.ts
Comment (tobitege) [reply frontend/src/services/fileService.ts:1]: I'll revise those changes, took'em as they were.

File: frontend/src/components/file-explorer/CodeEditor.tsx
Comment (tobitege) [reply frontend/src/components/file-explorer/CodeEditor.tsx:1]: The file itself came from an AI, I just made it work.
Idk if we can move the main event handling stuff and the "Save" button+logic somewhere else?

**Comment by tobitege** [comment]
Thanks so much @amanape for the great feedback!
I'll revise the execption handling further per your suggestions.
With regard to detecting file changes from "outside" in order to notify the LLM: that's another can of worms, tbh. 🤣 
At some point the user has to take some responsibility when editing files during a task run, either way.

**Comment by tobitege** [comment]
Final minor touches are in. This should be it for now, hopefully.

**Review by amanape — APPROVED** [review]
LGTM!

**Comment by iFurySt** [comment]
Xnip2024-07-02_14-34-14
Xnip2024-07-02_14-34-22
Xnip2024-07-02_14-34-29

I think we should maintain a consistent UI style.🥶

**Comment by tobitege** [comment]
> I think we should maintain a consistent UI style.🥶

I think the idea was to differentiate between "unsaved" and "already saved" as a reminder. 🤔

File: frontend/src/services/fileService.ts
Comment (iFurySt) [inline frontend/src/services/fileService.ts:8]: const encodedFile = encodeURIComponent(file);

this can be kept

File: frontend/src/services/fileService.ts
Comment (iFurySt) [inline frontend/src/services/fileService.ts:70]: why don't we use the `try...catch` pattern?

**Comment by iFurySt** [comment]
> I think the idea was to differentiate between "unsaved" and "already saved" as a reminder. 🤔

This is good. I just think the button color and the toast color stand out a bit too much.😅

File: frontend/src/services/fileService.ts
Comment (tobitege) [reply frontend/src/services/fileService.ts:70]: That's because @amanape had recommended to let the exceptions be caught at a higher level, not here.
But I see that I missed to re-add the encodedURIComponent usage. 😢 
Will do that in a minute.

**Comment by tobitege** [comment]
@iFurySt re-added the one encoding location.

Re: toast colors: those are the defaults, I think

**Comment by tobitege** [comment]
Thanks for your time and feedback, @iFurySt ! 🤗

**Comment by assertion** [comment]
> Thanks so much @amanape for the great feedback! I'll revise the execption handling further per your suggestions. With regard to detecting file changes from "outside" in order to notify the LLM: that's another can of worms, tbh. 🤣 At some point the user has to take some responsibility when editing files during a task run, either way.

@tobitege This is an important point, but I respectfully have a different viewpoint. While user responsibility is crucial, I believe we should implement a mechanism to track modifications made during a task run. For instance, we could add these changes to the history so that the LLM is aware of any edits the user makes. This way, we ensure that the LLM has the most up-to-date context, preventing potential issues like creating duplicate files, calling renamed functions incorrectly, or updating outdated line ranges like @amanape mentioned.

**Comment by assertion** [comment]
> Admittedly, my statement was a bit too flippant and I basically agree with you, that if there's a practical way to make this work without being too overcomplicated, we should look into this as a feature.
> 
> A while back I took a look at one potential rough way to do this, but discarded it since it might require an extra service. Something along the line of this to assist in the agent skills:
> 
> ```python
> class FileChangeHandler(FileSystemEventHandler):
>     def __init__(self):
>         self.modified_files = set()
> 
>     def on_modified(self, event):
>         if event.is_directory:
>             return
>         self.modified_files.add(event.src_path)
> 
>     def clear_modified_files(self):
>         self.modified_files.clear()
> 
> file_change_handler = FileChangeHandler()
> observer = Observer()
> observer.schedule(file_change_handler, path='.', recursive=True)
> observer.start()
> 
> def has_file_changed(file_path):
>     changed = file_path in file_change_handler.modified_files
>     file_change_handler.clear_modified_files()
>     return changed
> ```

Thanks for reply @tobitege , I create a issue to track this, hope to get more input from the community.

**Pull request merged**
